Add v1 of the SDK

nvtaveras opened the pull request:
### Description

This adds the v1 of the SDK for interacting with the Multi-Collateral Mento contracts. Most of the time I spent on this was trying to come up with a simple interface so that consumers don't need to know much about the internals of our contracts.

The SDK is built under two assumptions:

1. Each exchange has exactly 2 assets
2. There's only one exchange for a given pair of assets

While I know both of these assumptions won't be true in the future I thought it didn't make a lot of sense to overcomplicate things at the moment when we won't have them at launch. Nonetheless the code should be structured in a way which should make extending to more assets and exchanges relatively easy.

From a consumer perspective the usage will look like this:
```typescript
const mento = await mento.create(ethersProvider, ethersSigner)

// this is an expensive call since it also fetches symbols for the assets
// but it can be cached by consumers
const pairs = await mento.getTradeablePairs()
//  [
        [{
          address: '0x62492A644A588FD904270BeD06ad52B9abfEA1aE',
          symbol: 'cUSD'
        },
        {
          address: '0xdDc9bE57f553fe75752D61606B94CBD7e0264eF8',
          symbol: 'CELO'
        }],
      ...
    ]
    

// selling 1 CELO for cUSD
const one = ethers.utils.parseUnits('1', 18)
const amontOutMinQuote = await mento.getAmountOut(
    celoAddr,
    cUSDAddr,
    one
)

// currently not added because I ran into some issues with this while writing e2e tests, but soon
const allowanceTx = mento.increaseTradingAllowance(celoAddr, one)

const tx = await mento.swapIn(
    celoAddr,
    cUSDAddr,
    one,
    amontOutMinQuote
)
```

And the idea would be that once we have exchanges with 2+ assets or multiple exchanges with different price functions for the same set of assets we can keep the same interface and internally do the job of fetching the correct exchanges and the best quote without passing that burden to the consumer.

### Tested

Everything is unit tested under test/
```
$ dts test --coverage
 PASS  test/util.test.ts
 PASS  test/mento.test.ts
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
----------|---------|----------|---------|---------|-------------------
All files |     100 |      100 |     100 |     100 |
 mento.ts |     100 |      100 |     100 |     100 |
 utils.ts |     100 |      100 |     100 |     100 |
----------|---------|----------|---------|---------|-------------------

Test Suites: 2 passed, 2 total
Tests:       16 passed, 16 total
Snapshots:   0 total
Time:        6.125 s
Ran all test suites.
```

I'm working on adding some e2e tests that execute swaps on the different networks for all the exchanges but I ran into some issues related to increasing the allowance before trading. Will add those soon

### Related issues

- Fixes https://github.com/mento-protocol/mento-core/issues/15

### Documentation

No docs written on this yet, will probably create another issue about it to define an acceptance criteria for that. My current feeling is that it makes sense to have the docs in gitbook and link to them in the readme here but I'm curious to hear others opinion.

jmrossy commented on LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Just want to confirm this is the desired license since most things used to use Apache 2 at cLabs. 
Noticed the core repo is missing a license file so would be good to fix that too: https://github.com/mento-protocol/mento-core/

bowd replied:
I added a not to our backlog to get licences in order across all repos, we can merge this with MIT for now, as the decision feels a bit outside of the scope.

jmrossy commented on README.md:
````diff
@@ -1,22 +1,7 @@
-# Run all tests with yarn
-yarn test
-```
+- [ ] TODO: Adds docs link
````
Note, the contents of this Readme will be what's shown on NPM when you publish, so info oriented more to SDK users may be a good idea. Feel free to ignore if this is temporary.

bowd replied:
Yeah let's put in at least a bit of work on the readme to make it a bit more friendly to SDK consumers.

nvtaveras replied:
Chatted with Bogdan about this, we'll create a separate issue to write the docs and add them to https://docs.mento.org/ and will link it here afterwards

jmrossy commented on package.json:
```diff
@@ -1,32 +1,65 @@
-  "version": "1.0.0",
-  "description": "The SDK for Mento Protocol",
-  "main": "index.js",
+  "name": "@mentoprotocol/sdk",
```
I think the core package uses `@mento-protocol` with a dash

bowd replied:
Yeah and also, in the other packages we've used the repo name for the package as well, we should be consistent. Let's for now rename those to `@mento-protocol/mento-sdk` to be in line with the other packages, and decide about all of them if we want to change.

nvtaveras replied:
Updated, but noticed that the package in `mento-core` is named `@mento-protocol/core-contracts`. Should we update that as well?

jmrossy commented on package.json:
```diff
@@ -1,32 +1,65 @@
+  "devDependencies": {
+    "@size-limit/preset-small-lib": "^8.1.0",
+    "@tsconfig/recommended": "^1.0.1",
+    "dts-cli": "^1.6.0",
```
Haven't heard of dts before. It seems nice but not very popular. Consider that the building/bundler will have full control of the outputted code and it's nearly impossible to detect malicious payloads a rogue tool like this can sneak in. For that reason, I'd recommend avoiding this and configuring TS manually. Depending on who the consumers of this are, you probably don't even need a bundler, just the TS output is probably good enough! I can help with that if you need

bowd replied:
Agreed. Let's keep it lean. We can merge this PR like this and revert the change in another ticket.

nvtaveras replied:
Great point on malicious payloads, will keep it like this for now as Bogdan already created a separate ticket and do it there

jmrossy commented on package.json:
```diff
@@ -1,32 +1,65 @@
-  "license": "ISC",
-  "bugs": {
-    "url": "https://github.com/mento-protocol/mento-sdk/issues"
+  "prettier": {
```
Nit, consider moving this to a prettier config file, as is the more common convention

nvtaveras replied:
@bayological I believe the repo already had this and you moved it here when setting up dts. Any particular reason or should we just move it to a config file?

bayological replied:
No, this is just how it came with the template. I'm for moving to a separate file.

jmrossy commented on package.json:
```diff
@@ -1,32 +1,65 @@
+  "typings": "dist/index.d.ts",
+  "files": [
+    "dist",
+    "src"
```
src should not be included. This specifies which files should be included in the bundle published to npm

jmrossy commented on package.json:
```diff
@@ -1,32 +1,65 @@
+  "license": "MIT",
+  "author": "Mento Labs",
+  "main": "dist/index.js",
+  "module": "dist/sdk.esm.js",
```
should this be index.esm.js?

jmrossy commented on package.json:
```diff
@@ -1,32 +1,65 @@
+  "jest": {
+    "testEnvironment": "node"
+  },
+  "peerDependencies": {},
```
Nit, not needed if empty

jmrossy commented on package.json:
```diff
@@ -1,32 +1,65 @@
-    "husky": "^8.0.1",
-    "lint-staged": "^13.0.3",
-    "prettier": "^2.7.1"
+    "@ethersproject/providers": "^5.7.2",
```
Not needed as the ethers dep below already includes it

nvtaveras replied:
good catch, I think this was just carried over from when first setting up stuff

jmrossy commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+  IExchangeProvider,
+  IExchangeProvider__factory,
+} from '@mento-protocol/mento-core-ts'
+import { Provider, TransactionResponse } from '@ethersproject/providers'
```
Instead of a separate import for this, a common way would be to import `providers` from ethers and then use `providers.TransactionResponse`, etc.

jmrossy commented on package.json:
```diff
@@ -1,32 +1,65 @@
+    "dts-cli": "^1.6.0",
+    "husky": "^8.0.2",
+    "size-limit": "^8.1.0",
+    "tslib": "^2.4.1",
```
What is this needed for?

nvtaveras replied:
seems no needed indeed, removing

jmrossy commented on src/index.ts:
```diff
@@ -0,0 +1,4 @@
+/* istanbul ignore file */
```
Is this needed if the project is using jest?

nvtaveras replied:
this was me being nit to exclude `index.js` from the coverage report 😅

jmrossy commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+import { strict as assert } from 'assert'
+
+interface Exchange {
+  providerAddr: string
```
Optional tip: Define a global `Address` type to be string to make it more clear when you mean for a string type be to an address. Example: https://github.com/hyperlane-xyz/hyperlane-warp-ui-template/blob/main/src/global.d.ts

nvtaveras replied:
I like that!

jmrossy commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+   * @param ethersSigner an optional ethers signer to execute swaps (must be connected to a provider)
+   * @returns a new Mento object instance
+   */
+  static async create(ethersProvider: Provider, ethersSigner?: Signer) {
```
Optional: it's somewhat common to see `providerOrSigner: Provider | Signer` in methods like this, since often Signers already have connected providers in them

jmrossy commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+    for (let exchange of exchanges) {
+      const asset0 = exchange.assets[0]
+      const asset1 = exchange.assets[1]
+      let symbols = await Promise.all([
```
symbols could be `const`, recommend setting up lint and/or ts rules to enforce that. 
A reasonable ruleset example here: https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/.eslintrc

nvtaveras replied:
Agree, would just fix it here right now and create an issue for the linting rules

jmrossy commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+      amountIn,
+      amountOutMin
+    )
+    return await this.signer.sendTransaction(tx)
```
It's good for the SDK to have methods that send transactions for you, but I think a common use case will be for DApps to manage the sending themselves. They may need control of that to interface well with whatever libs their using (e.g. Web3Modal, Rainbowkit) and also for finer grained control of progress reporting. And maybe for gas configuration too.

So I'd recommend either a `sendTransaction` argument flag or a separate method so users can get the populated tx back.

bowd replied:
+1 we we should always return transaction objects, not execute them. I don't even think we need the `sendTransaction` flag tbh. I think 99% of the time consumers will want to have control of the sending and will have a pipeline to handle that.

jmrossy commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+      amountIn,
+      amountOutMin
+    )
+    return await this.signer.sendTransaction(tx)
```
Unrelated, `return await` has some subtle pitfalls in javascript and you generally don't want to do it. Either `return this.signer.sendTransaction` directly which will return a promise callers can await for await it then return a value on a separate line.

jmrossy commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+      amountOut,
+      amountInMax
+    )
+    return await this.signer.sendTransaction(tx)
```
ditto as above regarding the sending and also the `return await`

jmrossy commented on src/utils.ts:
```diff
@@ -0,0 +1,43 @@
+import { Contract, constants } from 'ethers'
+
+import { Provider } from '@ethersproject/providers'
```
ditto

jmrossy commented on src/utils.ts:
```diff
@@ -0,0 +1,43 @@
+  provider: Provider
+): Promise<string> {
+  const celoRegistryAddress = '0x000000000000000000000000000000000000ce10'
+  const brokerIdentifier = 'Broker'
+
+  const registryAbi = [
+    'function getAddressForString(string calldata identifier) external view returns (address)',
+  ]
```
consider moving these to a `consts` file

nvtaveras replied:
I generally use `consts` files for constants that are either reused or when I have several that can be grouped. Otherwise I prefer keeping them in the place they are used. Happy to change it others have a similar opinion here though

jmrossy commented on src/utils.ts:
```diff
@@ -0,0 +1,43 @@
+  const contract = new Contract(celoRegistryAddress, registryAbi, provider)
+
+  const brokerAddress = await contract.getAddressForString(brokerIdentifier)
+  if (brokerAddress === constants.AddressZero) {
```
consider also checking if broker is not a valid address to cover more cases here.
e.g.
`if (!isValidAddress(brokerAddress) || brokerAddress === constants.AddressZero)`

Relevant: https://github.com/mento-protocol/mento-web/blob/main/src/utils/addresses.ts#L4

nvtaveras replied:
Hmm I don't think this validation would add much value here given that the address is read from the on-chain registry contract. If it was somehow set to an invalid address the whole deployment would be broken. wdyt?

jmrossy commented on test/mento.test.ts:
```diff
@@ -0,0 +1,342 @@
+import { Contract, utils } from 'ethers'
```
Totally optional: co-locating the tests with their files in the same folder is more common in JS these days and can help keep things organized as projects grow.

bowd replied:
Yeah I also liked that +1

jmrossy commented on tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "extends": "@tsconfig/recommended/tsconfig.json",
+  "include": ["src", "types"],
```
Is there a `types` folder?

nvtaveras replied:
nope, removing

jmrossy commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+   */
+  static async create(ethersProvider: Provider, ethersSigner?: Signer) {
+    return new Mento(
+      await getBrokerAddressFromRegistry(ethersProvider),
```
Another passing though: it may be nice to have a synchronous option for instantiating a Mento class. So maybe another static method like `createWithRegistry` may be nice

nvtaveras replied:
Not sure exactly what did you meant by `createWithRegistry`. 

Would it be more like `createWithAddress(brokerAddress, providerOrSigner)` where the consumer already knows the address of the underlying broker they want to interact with?

bowd commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,32 @@
+        with:
+          node-version: ${{ matrix.node }}
+
+      - name: Install deps and build (with cache)
+        uses: bahmutov/npm-install@v1
```
+1 we should use a `yarn` action.

bowd reviewed (changes requested):
Great job on the first version of the SDK!
There are a few things to address, and some I've extracted into new issues as per the comments.

philbow61 reviewed (approved):
LGTM 😎 
enjoyed the refactoring of the tests 💪

nvtaveras commented on src/mento.ts:
```diff
@@ -0,0 +1,260 @@
+   * Returns a list of all the pairs that can be traded on Mento
+   * @returns The list of tradeable pairs in the form of [{address, symbol}]
+   */
+  async getTradeablePairs(): Promise<[Asset, Asset][]> {
```
that's a great question, I'm also not sure how this should be approached. 

I believe ethers will throw an early error though when trying to estimate the gas for the tx (so it wouldn't require the user submitting the tx and finding out after that trading was disabled). 

I'll create an issue for this so we can think more deeply about how we want to approach

nvtaveras commented:
@jmrossy @bowd Updated addressing all comments except for the `providerOrSigner` recommendation. 

I like this idea but I'm playing around with it locally before making the changes. I'm mainly thinking if we want to just change the create function signature and then internally handle signer/provider separately as we do right now or if we want to use a single signerOrProvider object everywhere, which seems like it should work with ethers and the typechain types.

jmrossy reviewed (approved):
Thanks for addressing my comments! Of the outstanding issues, I think the use of `dts` is the most high-pri but open to circling back on that in another PR :)

nvtaveras commented:
Pushed the last v1 changes 🚢 

- The `signerOrProvider` interface is now used everywhere (+ tests)
- Added the `increaseTradingAllowance` function in the Mento class
- Added an e2e sample usage to the readme in the meantime while we update the docs

bowd commented on README.md:
```diff
@@ -1,22 +1,61 @@
+  tokenIn,
+  tokenOut,
+  // 95% of amountIn to allow some slippage
+  utils.parseUnits((one * 0.95).toString(), 18)
```
While in some sense this achieves the same result, I feel like it would be more natural to scale the return value of getAmountOut by 0.95 instead of the amountIn passed to it. Maybe call the result `expectedAmountOut` and then defined `minAmountOut = expectedAmountOut * 0.95`

nvtaveras replied:
Yeah this was my first thought as well but BigNumber in ethers only supports integers, so I can't do `expectedAmountOut.mul(0.95)`

nvtaveras replied:
chatted offline, changed this to `minAmountOut = expectedAmountOut.mul(95).div(100)`

bowd commented on src/mento.ts:
```diff
@@ -0,0 +1,322 @@
+  ): Promise<providers.TransactionRequest> {
+    if (!Signer.isSigner(this.signerOrProvider)) {
+      throw new Error(
+        'A signer is required to increase the populate the increaseAllowance tx object'
```
typo in sentence here

bowd commented on src/mento.ts:
```diff
@@ -0,0 +1,322 @@
+    )
+
+    // The contract call doesn't populate all of the signer fields, so we need an extra call for the signer
+    return this.broker.signer.populateTransaction(tx)
```
Feels more natural to use `this.signerOrProvider` because that's what you assert in the beginning of the function.

bowd commented on src/mento.ts:
```diff
@@ -0,0 +1,322 @@
+    )
+
+    // The broker's call doesn't populate all of the signer fields, so we need an extra call for the signer
+    return this.broker.signer.populateTransaction(tx)
```
ditto

bowd commented on src/mento.ts:
```diff
@@ -0,0 +1,322 @@
+    )
+
+    // The broker's call doesn't populate all of the signer fields, so we need an extra call for the signer
+    return this.broker.signer.populateTransaction(tx)
```
dittox2

bowd reviewed (approved):
👍 I think this is good to go as a base and we should make sure we've captured next steps

nvtaveras commented:
@bowd updated
